flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java

```diff
@@ -255,4 +257,33 @@ private Utils() {
 		}
 		return result;
 	}
+
+	/**
+	 * Recursive directory copy to work around FileSystem implementations that do not implement it.
+	 * @param fs
+	 * @param localPath
+	 * @param remotePath
+	 * @throws IOException
+     */
+	protected static void copyFromLocalFile(final FileSystem fs, final Path localPath, final Path remotePath) throws IOException {
+		File localFile = new File(localPath.toUri());
+		if (localFile.isDirectory()) {
+			for (File file : localFile.listFiles()) {
+				copyFromLocalFile(fs, new Path("file://" + file.getAbsolutePath()), new Path(remotePath,file.getName()));
+			}
+		} else {
+			fs.copyFromLocalFile(localPath,remotePath);
+		}
+	}
+
+	public static Path checkScheme(final Path localRsrcPath) throws IOException {
```

Comment:
Yes on changes.  The short answer to what will happen if scheme is undefined is it will fail. I don't remember offhand exactly why, but I think may have been because the default file system is S3a, which I suppose would become the default fs for unqualified paths. I noticed that all but one or two callers' paths do have qualified scheme. Will add comment.